accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,498 @@
+
+Note: Portable apps are not a meaningful option for client apps since almost all client apps require an executable launcher, and there is only ever one generated. Portable apps are only meaningful for commandline apps.
+
+Note: In theory, we could consider an option to create multiple apphosts for different RIDs, either of different names or in different directories. Again, this capability would need to be opt-in.
+
+### Rationalize RID folders for RID-specific and portable apps
+
+There have been some issues filed (see appendix) that suggest that managed libraries are not resilient to RID-specific libraries changing location as a result of changing build type (portable to RID-specific). This would also be true for single file deployment builds.
+
+We could enable (as an option) creating `runtimes` folders for RID-specific apps to make the folder layout the same across deployment types. Doing that would put single file at a significant disadvantage. We'd rather encourage the ecosystem to support both folder layouts. If we had to choose one of these deployments types to disadvantage, it would be portable (with the `runtimes` folders), not RID-specific, which is the likely the opposite of what developers are tuning for today.
+
+We need to invest more in docs and tools to the end of clear and actionable guidance.
+
+### Use SDK RID as default RID
+
+RID-specific apps should default to the SDK RID as an implicit RID, and no longer require a RID to be specified. The SDK already does this for the apphost for portable apps. This proposal is a straightforward extension of that behavior.
+
+This means that the following commands would produce assets in terms of the implicit RID, such as with the following commands.
+
+- `dotnet publish --self-contained`
+- `dotnet publish /p:PublishSingleFile=true`
+
+Note: As already suggested, this change would make Dockerfiles MUCH easier to write. Needing to specify RIDs is probably the worst part of using .NET with Docker.
+
+Specifying a RID should have no effect on the build type/deployment model. It should simply refine the existing model. This change largely follows from using the SDK RID as the implicit default for all types of RID-specific apps. When a RID is specified, it is substituted for the implicit RID.
+
+This means that the following commands would produce a RID-specific framework-dependent app:
+
+- `dotnet build -r linux-x64`
+- `dotnet publish -r linux-x64 /p:PublishReadyToRun=true`
+
+Developer would add `--self-contained` or `--self-contained true` to create a self-contained app, like with the following command.
```

Comment:
Since `--self-contained` is kind of a lot to type. What do you think about adding a short option `-s` for this switch?

`dotnet publish -r linux-x64 -s`

Reply:
Yup. Fan of that (or something like it).

If/when we do that, I'd suggest that we enable an implicit RID by default. That would enable this scenario:

```console
PS C:\Users\rich\app> dotnet publish --self-contained
Microsoft (R) Build Engine version 16.8.3+39993bd9d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored C:\Users\rich\app\app.csproj (in 61 ms).
C:\Program Files\dotnet\sdk\5.0.103\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(136,5): error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false. [C:\Users\rich\app\app.csproj]
```

Reply:
I would have the RID default to the host OS's platform and architecture.

Reply:
Filed an issue on this. https://github.com/dotnet/sdk/issues/19576

Reply:
Perfect! Thanks.